src/textual/widget_navigation.py

```diff
@@ -0,0 +1,192 @@
+"""
+Utilities to move index-based selections backward/forward.
+
+These utilities concern themselves with selections where not all options are available,
+otherwise it would be enough to increment/decrement the index and use the operator `%`
+to implement wrapping.
+"""
+
+from __future__ import annotations
+
+from functools import partial
+from itertools import count
+from typing import TYPE_CHECKING, Literal, Protocol, Sequence
+
+from typing_extensions import TypeAlias, TypeVar
+
+if TYPE_CHECKING:
+    from .widget import Widget
+
+
+class Disableable(Protocol):
+    """Non-widgets that have an enabled/disabled status."""
+
+    disabled: bool
+
+
+Direction: TypeAlias = Literal[-1, 1]
+"""Valid values to determine navigation direction.
+
+In a vertical setting, 1 points down and -1 points up.
+In a horizontal setting, 1 points right and -1 points left.
+"""
+_D = TypeVar("_D", bound=Disableable)
+_W = TypeVar("_W", bound="Widget")
+
+
+def distance(index: int, start: int, direction: Direction, wrap_at: int) -> int:
```

Comment:
Should really start with a very. At least `get_distance`, but can we come up with a more descriptive name?